docs/design/2022-03-31-concurrent-ddl-framawork.md

```diff
@@ -0,0 +1,233 @@
+| table_id       | bigint(20) | YES  |      | The table ID relate to this DDL  |
+| job_meta       | blob       | YES  |      | The arguments of this DDL job    |
+| is_drop_schema | bigint(20) | YES  |      | True if the DDL is a drop schema |
++----------------+------------+------+------+---------+------------------------+
+```
+
+Table `mysql.tidb_ddl_reorg` contains the reorg job data.
+```
++---------------+------------+------+------+
+| Field         | Type       | Null | Key  |
++---------------+------------+------+------+
+| job_id        | bigint(20) | NO   |      |
+| ele_id        | bigint(20) | YES  |      |
+| curr_ele_id   | bigint(20) | YES  |      |
+| curr_ele_type | blob       | YES  |      |
+| start_key     | blob       | YES  |      |
+| end_key       | blob       | YES  |      |
+| physical_id   | bigint(20) | YES  |      |
++---------------+------------+------+------+
+```
+
+Table `mysql.tidb_ddl_history` stores the finished DDL job.
+```
++---------------+------------+------+------+
+| Field         | Type       | Null | Key  |
++---------------+------------+------+------+
+| job_id        | bigint(20) | NO   | PRI  |
+| job_meta      | blob       | YES  |      |
++---------------+------------+------+------+
+```
+
+In bootstrap step, TiDB will build these tables meta and put it into tikv directly. For new cluster, TiDB will also build `mysql` schema meta.
+
+### DDL operations
+
+Use the following DDL operations to manage the DDL jobs:
+```sql
+insert into mysql.tidb_ddl_job values (...) -- add ddl jobs
+delete from mysql.tidb_ddl_job where job_id = ... -- delete ddl jobs
+update mysql.tidb_ddl_job set job_meta = ... where job_id = ... -- update ddl jobs
+select * from mysql.tidb_ddl_job -- get ddl jobs
+```
+
+### DDL job manager
+
+DDL job manager will find the runnable DDL job and dispatch the jobs to DDL workers.
+```golang
+for {
+    if !isOwner {
+        // sleep a while
+        continue
+    }
+
+    // if there are job coming
+    select {
+        case <-ddlJobCh:
+        case <-ticker.C:
+        case <-notifyDDLJobByEtcdCh:
+    }
+
+    job = findrunnableJob()
+    if job != nil {
+        waitForFreeWorker();
+        go runJob(job)
+    }
+}
+```
+
+To prevent all the workers in worker pool occupied by the long run job like `add index`, TiDB will divide the pool into two types: `addIndex` worker pool and `general` worker pool, then the DDL job manager will be
+
+```golang
+for {
+    // ...
+    generalJob = findRunnableGeneralJob()
+    reorgJob = findRunnableReorgJob()
+    if generalJob != nil {
+        waitForFreeGeneralWorker();
+        go runGeneralJob(job)
+    }
+    if reorgJob != nil {
+        waitForFreeReorgWorker();
+        go runReorgJob(job)
+    }
+}
+```
+
+The runnable job defines as
+1. not running.
+2. the minimum job id on the same table.
+3. no smaller `drop schema` job id on the same schema.
+4. no smaller job id on the same schema if the job type is `drop schema`.
```

Comment:
I remember apart from `drop schema`, some implicit table properties (like COLLATION) are still refer to schema properties (not sure what's our implementation, will check later). so maybe rule 3 can be expressed by the dependency relation? And rule 4 as well.

Reply:
For 'COLLATION' and 'PLACEMENT RULES' attribute of schema, they are 'default values' of a schema, that is, changing those attributes of a schema does not affect existing tables in that schema. So dependency is not a problem.

Reply:
For 'COLLATION' and 'PLACEMENT RULES' attribute of schema, they are 'default values' of a schema, that is, changing those attributes of a schema does not affect existing tables in that schema. So dependency is not a problem.

Reply:
it's not about existing tables, the race happens when CREATE TABLE and ALTER DATABASE. For example

DDL1: ALTER DATABASE db COLLATE <new_collation>
DDL2: CREATE TABLE db.tbl

if DDL2 is sent to TiDB after DDL1, we should make sure db.tbl is created with <new_collation>. However by current rule DDL2 is concurrent with DDL1, so DDL2 may be executed before DDL1 thus it uses <old_collation>

Reply:
Good point, I think it's reasonable.